### PR TITLE
Upgrade PouchDB support to v6.1.2 (applies to all pouchdb-* packages) 

### PR DESCRIPTION
**Pull request by spaulg**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://pouchdb.com/api.html
- [x] Increase the version number in the header if appropriate.
- [x] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "../tslint.json" }`.

This change adds support for the current (latest) version of PouchDB (v6.1.2 as of this PR, v6.2.0 is still pending).

The change applies to multiple packages, but for the most part of version updates to confirm support for the newer versions, are minimal and under the "pouchdb-*" prefix.

I have not copied the original file to a new file with the version number - I can do this if required.

### CODE REVIEW

**Comment by AGBrown** [comment]
@spaulg - it might be an idea to take me off the author lists now as DT bot relies on me for a +1 on each PR, but this has all moved on a lot since I worked on it so I can't add anything in PR reviews anymore.

File: types/pouchdb-core/index.d.ts [hunk 1]
```diff
@@ -55,6 +55,15 @@ declare namespace PouchDB {
             rev: RevisionId;
             status: Availability;
         }
+        interface RevisionDiffList {
+            [documentId : string]: string[];
+        }
+        interface RevisionDiff {
+            [change : string]: string[];
```
Comment (geppy) [inline types/pouchdb-core/index.d.ts]: My understanding is that this corresponds to [this example response](https://pouchdb.com/api.html#example-response-20) from the docs, and that the key tells you which *kind* of change you're dealing with.

@nolanlawson Is there a fixed set of keys here (e.g. "missing")?  If so we can use an enum / union type.

File: types/pouchdb-core/index.d.ts [hunk 2]
```diff
@@ -55,6 +55,15 @@ declare namespace PouchDB {
             rev: RevisionId;
             status: Availability;
         }
+        interface RevisionDiffList {
+            [documentId : string]: string[];
```
Comment (geppy) [inline types/pouchdb-core/index.d.ts]: The compiler doesn't care, but to help out users could you type `documentId` as a `DocumentId`?

File: types/pouchdb-core/index.d.ts [hunk 3]
```diff
@@ -55,6 +55,15 @@ declare namespace PouchDB {
...
+        }
+        interface RevisionDiff {
+            [change : string]: string[];
+        }
+        interface RevisionDiffResponse {
+            [documentId : string]: RevisionDiff
```
Comment (geppy) [inline types/pouchdb-core/index.d.ts]: Same as above, can you type `documentId` as `DocumentId`?

**Review by geppy — COMMENTED** [review]
This looks great, thanks!  I've suggested some minor rewordings.

File: types/pouchdb-core/index.d.ts (hunk 1 above)
Comment (spaulg) [inline types/pouchdb-core/index.d.ts]: Thanks for reviewing @geppy.

Looking at the PouchDB source, it just fires an xhr off to CouchDB and returns whatever the response is via the promise/callback. The [CouchDB docs](http://docs.couchdb.org/en/2.0.0/api/database/misc.html#db-revs-diff) document the response key as either "missing" or "possible_ancestors".

That aside, I've tried putting an enum together, however my IDE is flagging the code with an error: "TS1023: An index signature parameter type must be string or number". 

```
enum RevisionDiffStatus {'missing', 'possible_ancestors'}

interface RevisionDiff {
    [Change : RevisionDiffStatus]: string[];
}
```

Looking this error up, it doesn't look like Typescript supports enumerations as index types.

I'm somewhat new to TypeScript however and may be doing something wrong. Could you advise how it should be accomplished?

Comment (spaulg) [inline types/pouchdb-core/index.d.ts]: Specifically, see https://github.com/Microsoft/TypeScript/issues/2491 & https://github.com/Microsoft/TypeScript/issues/13042

Comment (fredgalvao) [inline types/pouchdb-core/index.d.ts]: Can you try using literal join types? Like so:
```ts
type RevisionDiffStatus = 'missing' | 'possible_ancestors'
```

That, imo, should trick typescript into believing that type is a subtype of String (which it is), and thus allow it.

Comment (fredgalvao) [inline types/pouchdb-core/index.d.ts]: Nop, just tried it, the compiler doesn't even try to traverse that type if it differs from `number | string`.

Comment (fredgalvao) [inline types/pouchdb-core/index.d.ts]: At least we can do
```ts
interface RevisionDiff {
    missing?: string[];
    possible_ancestors?: string[];
}
```
if we indeed know those are the possible outcomes. It's not awesome, but it works.

Comment (geppy) [inline types/pouchdb-core/index.d.ts]: Awesome, thanks for trying this!  (And for teaching me that that doesn't work!)

Comment (geppy) [inline types/pouchdb-core/index.d.ts]: Another option would be this:

```typescript
interface RevisionDiff {
    missing?: string[];
    possible_ancestors?: string[];
    [reason: string]: string[];
}```

@spaulg Whatever you want to go with works for me.

Comment (fredgalvao) [inline types/pouchdb-core/index.d.ts]: That allows for this, which I wouldn't want to be accepted:
```ts
let a: RevisionDiff;
console.log(a.posible_ancestors); //<-- typo gone rogue through the system
```

Comment (spaulg) [inline types/pouchdb-core/index.d.ts]: I have now updated the code to match this implementation:

```typescript
interface RevisionDiff {
    missing?: string[];
    possible_ancestors?: string[];
}
```

**Pull request merged**